rust/actyx/ax-aql/src/language/type_check.rs

```diff
@@ -0,0 +1,575 @@
+use super::workflow::{Participant, Workflow, WorkflowStep};
+use crate::{Ident, Ind, Index, Label, NonEmptyVec, Num, Query, SimpleExpr, Type, TypeAtom};
+use std::{collections::BTreeSet, ops::Deref, sync::Arc};
+pub(crate) struct QueryTypeCheckError<'a> {
+    pub(crate) span: pest::Span<'a>,
+    pub(crate) error_type: QueryTypeCheckErrorType,
+}
+
+type ExtraMessage = String;
+pub(crate) enum QueryTypeCheckErrorType {
+    UndeclaredEvent,
+    UndeclaredParticipant,
+    UndeclaredWorkflowCall,
+    InvalidIndex(ExtraMessage),
+}
+struct WorkflowAnalysisResource<'a> {
+    workflow: &'a Workflow<'a>,
+    participants_set: BTreeSet<&'a Ident>,
+}
+
+impl<'a> WorkflowAnalysisResource<'a> {
+    fn from(workflow: &'a Workflow) -> Self {
+        Self {
+            workflow,
+            participants_set: workflow
+                .args
+                .iter()
+                .map(|x| {
+                    let participant: &Participant = x;
+                    match participant {
+                        Participant::Role(ident) => ident,
+                        Participant::Unique(ident) => ident,
+                    }
+                })
+                .collect(),
+        }
+    }
+}
+
+pub(crate) struct QueryTypeCheck<'a> {
```

Comment:
Overall, type checking is a function, not an object, so I’d start from functions, not methods. If the argument lists become too long and repetitive, then some arguments can be passed within a single struct.